File: 120-triangle/ken/120.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """

        current = [0 for i in xrange(len(triangle))]

        for row in triangle:
            for i in xrange(len(row) - 1, -1, -1):
                prev = 0
                if i == 0:
                    prev = current[i]
                elif i == len(row) - 1:
                    prev = current[i - 1]
                else:
                    prev = min(current[i - 1], current[i])

                current[i] = prev + row[i]

        return min(current)

File: 120-triangle/leochang/120.cpp
class Solution {
public:
	int minimumTotal(vector<vector<int>>& triangle) {
		int rows = triangle.size();
		vector<int> minPathSum(triangle.back());
		for (int i = rows - 2; i >= 0; i--) {
			for (int j = 0; j <= i; j++) {
				minPathSum[j] = min(minPathSum[j], minPathSum[j + 1]) +triangle[i][j];
			}
		}
		return minPathSum[0];
	}
};
